subsys/net/lib/lwm2m/lwm2m_message_handling.c

```diff
@@ -2472,18 +2522,68 @@ static int do_send_op(struct lwm2m_message *msg, uint16_t content_format,
 	}
 }
 
+static bool lwm2m_timeseries_data_rebuild(struct lwm2m_message *msg, int error_code)
+{
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
+	struct lwm2m_cache_read_info *cache_temp;
+
+	if (error_code != -ENOMEM) {
+		return false;
+	}
+
+	cache_temp = msg->cache_info;
+
+	if (!cache_temp || !cache_temp->entry_size) {
+		return false;
+	}
+
+	/* Put Ring buffer back to original */
+	for (int i = 0; i < cache_temp->entry_size; i++) {
+		cache_temp->read_info[i].cache_data->rb.get_head =
+			cache_temp->read_info[i].original_get_head;
+		cache_temp->read_info[i].cache_data->rb.get_tail =
+			cache_temp->read_info[i].original_get_tail;
+		cache_temp->read_info[i].cache_data->rb.get_base =
+			cache_temp->read_info[i].original_get_base;
+	}
```

Comment:
Do we have the ring buffer locked while we do this?
Otherwise the head might go forward.

Reply:
Good Point. I need to change lwm2m_engine_send() that whole process is secured.